@hts-forking/src/slotmap.js

```diff
@@ -0,0 +1,158 @@
+/*-
+ * Hedera Hardhat Forking Plugin
+ *
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const { strict: assert } = require('assert');
+const { keccak256 } = require('ethers');
+
+const { toIntHex256, toSnakeCase } = require('./utils');
+const {
+    storageLayout: { storage, types },
+} = require('../resources/HtsSystemContract.json');
+
+class SlotMap {
+    constructor() {
+        /** @type {Map<bigint, {value: string, path: string, type: string}>} */
+        this._map = new Map();
+    }
+
+    /**
+     *
+     * @param {bigint} slot
+     * @param {string} value
+     * @param {string} path
+     * @param {string} type
+     */
+    store(slot, value, path, type) {
+        const prev = this._map.get(slot);
+        if (prev !== undefined)
+            throw new Error(`Slot \`${slot}\` in use by ${JSON.stringify(prev)}: ${value}@${path}`);
+        this._map.set(slot, { value: '0x' + value, path, type });
+    }
+}
+
+/**
+ * @type {{[t_name: string]: (value: string) => string[]}}
+ */
+const _types = {
+    t_string_storage: function (str) {
+        return this['t_bytes_storage'](Buffer.from(str).toString('hex'));
+    },
+    t_bytes_storage: hexstr => {
+        const len = Buffer.from(hexstr, 'hex').length;
+        const [header, lenByte, chunks] =
+            len > 31
+                ? [
+                      '0',
+                      len * 2 + 1,
+                      [...Array(Math.ceil(len / 32)).keys()].map(i =>
+                          hexstr.substring(i * 64, (i + 1) * 64).padEnd(64, '0')
+                      ),
+                  ]
+                : [hexstr, len * 2, []];
+
+        return [`${header.padEnd(64 - 2, '0')}${lenByte.toString(16).padStart(2, '0')}`, ...chunks];
+    },
+    t_uint8: val => [toIntHex256(val)],
+    t_uint256: val => [toIntHex256(val)],
+    t_int256: str => [toIntHex256(str ?? 0)],
+    t_address: str => [toIntHex256(str?.replace('0.0.', '') ?? 0)],
```

Comment:
🤔 , wouldn't it be better if we called mirror node to get the real EVM address of the account instead of relying on the long-zero address?

Reply:
Done.